# DATA ANALYSIS

You and your classmates have collected data from various locations about houses, including their features (e.g., area, number of rooms, condition) and prices (house prices and rents). The objective is to help a fictional real estate firm analyze trends and patterns in the housing market to make decisions about property investment and pricing strategies.

As the Data Scientist, your task is to analyze this data using measures of central tendency, dispersion, and correlation analysis. Based on your findings, you will provide insights that the firm can use to make informed decisions.

In [27]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [28]:
house_data=pd.read_excel('Housing-1.xlsx')

# Exploring the dataset

In [29]:
# Display the First 5 Rows
house_data.head()

Location  Area (Marla)  No of Rooms  No of Washrooms  No of Kitchens Garden  \
0  Chatter             6            4                2               1    Yes   
1  Chatter             8            6                3               1    Yes   
2  Chatter             7            5                3               1     No   
3  Chehlla             3            2                1               1     No   
4  Chehlla            10            6                3               1     No   

  Garage Balcony  Floors Condition   House Prices (Millions)  Rent Prices  \
0     No      No       1       Good                      9.0        20000   
1    Yes      No       1       Fair                     13.0        40000   
2    Yes      No       1       Fair                     12.0        20000   
3    Yes      No       1       Fair                      8.0        30000   
4    Yes     Yes       2       Good                     15.0        60000   

   Year of Construction  
0                2000.0  
1                2019.0  
2                1999.0  
3                   NaN  
4                2015.0

### Key insight 
We get a general sense of what the dataset contains like data structure /formate, column names, and sample values.


In [30]:
# Check Data Types and Missing Values
house_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74 entries, 0 to 73
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Location                 74 non-null     object 
 1   Area (Marla)             74 non-null     int64  
 2   No of Rooms              74 non-null     int64  
 3   No of Washrooms          74 non-null     int64  
 4   No of Kitchens           74 non-null     int64  
 5   Garden                   74 non-null     object 
 6   Garage                   74 non-null     object 
 7   Balcony                  74 non-null     object 
 8   Floors                   74 non-null     int64  
 9   Condition                72 non-null     object 
 10  House Prices (Millions)  73 non-null     float64
 11  Rent Prices              74 non-null     int64  
 12  Year of Construction     72 non-null     float64
dtypes: float64(2), int64(6), object(5)
memory usage: 7.6+ KB


### Key insight 
The dataset has 74 entries with 13 columns, including numerical and categorical data helps determine the type of analysis. Missing values exist in Condition, House Prices, and Year of Construction, requiring imputation. Key variables like Area, Rooms, and amenities will be critical for analyzing property trends and correlations.

In [31]:
house_data.describe()

Area (Marla)  No of Rooms  No of Washrooms  No of Kitchens     Floors  \
count     74.000000    74.000000        74.000000       74.000000  74.000000   
mean       8.608108     5.662162         3.094595        1.486486   1.581081   
std        3.877641     2.456621         1.536487        0.646238   0.573544   
min        3.000000     2.000000         1.000000        1.000000   1.000000   
25%        5.000000     4.000000         2.000000        1.000000   1.000000   
50%        8.000000     6.000000         3.000000        1.000000   2.000000   
75%       10.000000     7.000000         4.000000        2.000000   2.000000   
max       20.000000    15.000000         8.000000        4.000000   3.000000   

       House Prices (Millions)    Rent Prices  Year of Construction  
count                73.000000      74.000000             72.000000  
mean                 12.027397   30662.162162           2010.138889  
std                   5.479691   40488.888050             11.125792  
min                   6.000000    3000.000000           1970.000000  
25%                   8.000000   15500.000000           2006.750000  
50%                  10.000000   20000.000000           2013.500000  
75%                  15.000000   30000.000000           2018.000000  
max                  30.000000  300000.000000           2024.000000

### Key insight 
#### Missing Values:
House Prices: Only 73 entries exist out of 74, indicating 1 missing value.
Year of Construction: Only 72 entries exist, indicating 2 missing values.
#### Outliers:
Rooms: The maximum value is 15, which is unusually high compared to the mean (5.66).
Washrooms: The maximum is 8, higher than expected for most houses.
House Prices and Rent Prices: The wide range (6M–30M for prices and 3,000–30,000 for rent) indicates possible outliers.
#### Year of Construction:
The minimum year (1970) seems valid, but the maximum year (2024) could be future data that needs validation. 
These errors indicate the need for cleaning, such as handling missing values and checking for outliers.

## Handle Missing Data

In [32]:
house_data.isnull().sum()

Location                   0
Area (Marla)               0
No of Rooms                0
No of Washrooms            0
No of Kitchens             0
Garden                     0
Garage                     0
Balcony                    0
Floors                     0
Condition                  2
House Prices (Millions)    1
Rent Prices                0
Year of Construction       2
dtype: int64

### Key insight
Condition:  2 missing values. These can be filled with the mode (most common condition).                                      
House Prices (Millions):  1 missing value. It can be filled with the median or mean depending on the distribution.            
Year of Construction:  2 missing values. These can be filled with the median (most representative value) since it is numeric and likely has outliers.

In [33]:
# Fill Condition with mode
house_data['Condition '] = house_data['Condition '].fillna(house_data['Condition '].mode()[0])

# Fill House Prices with median
house_data['House Prices (Millions)'] = house_data['House Prices (Millions)'].fillna(house_data['House Prices (Millions)'].median())

# Fill Year of Construction with median
house_data['Year of Construction'] = house_data['Year of Construction'].fillna(house_data['Year of Construction'].median())

### Key insight
The missing values in the dataset were successfully handled by filling Condition with its mode, and both House Prices (Millions) and Year of Construction with their respective medians, ensuring data completeness without skewing distributions. This enables accurate analysis and reliable model training.

## Part 1: Central Tendency (Patterns in Data)

1 - What is the typical house price and rent price in the dataset? Decide whether the mean, median, or mode is the most appropriate measure of   central tendency for these variables, considering the presence of any outliers.

In [34]:
# Calculate statistics
house_mean = house_data['House Prices (Millions)'].mean()
house_median = house_data['House Prices (Millions)'].median()
house_mode = house_data['House Prices (Millions)'].mode()[0]  # Taking the first mode

rent_mean = house_data['Rent Prices'].mean()
rent_median = house_data['Rent Prices'].median()
rent_mode = house_data['Rent Prices'].mode()[0]  # Taking the first mode

# Print in the desired format
print(f"Analysis of House Prices and Rent Prices")
print(f" House Prices:")
print(f"  - Mean: {house_mean:.2f} million")
print(f"  - Median: {house_median:.2f} million (typical price)")
print(f"  - Mode: {house_mode:.2f} million")
print(f" Rent Prices:")
print(f"  - Mean: {rent_mean:.2f}")
print(f"  - Median: {rent_median:.2f} (typical price)")
print(f"  - Mode: {rent_mode}")


Analysis of House Prices and Rent Prices
 House Prices:
  - Mean: 12.00 million
  - Median: 10.00 million (typical price)
  - Mode: 9.00 million
 Rent Prices:
  - Mean: 30662.16
  - Median: 20000.00 (typical price)
  - Mode: 20000


### Key Insight:
The median is the most appropriate measure of central tendency for both house prices and rent prices due to potential skewness in the data, as it is less affected by extreme values or outliers. This provides a robust estimate of the typical price and rent in the dataset. Therefore, the typical house price is 10.0 million, and the typical rent price is 20,000.

2 - Compare the average number of rooms in houses across different locations. Which location tends to have the most spacious homes?

In [35]:
# Group by location and calculate the average number of rooms
average_rooms_per_location = house_data.groupby('Location')['No of Rooms'].mean()

# Find the location with the most spacious homes (highest average number of rooms)
most_spacious_location = average_rooms_per_location.idxmax()
most_spacious_rooms = average_rooms_per_location.max()

# Print the results
print("Average number of rooms in houses across different locations:")
print(average_rooms_per_location)

print(f"\nThe location with the most spacious homes is {most_spacious_location} with an average of {most_spacious_rooms} rooms.")


Average number of rooms in houses across different locations:
Location
Ambore             6.222222
Balapeer           3.000000
Bela Noor Shah     7.500000
Bela noor shah     7.000000
Bella Noor Shah    4.000000
Chatter            5.300000
Chehlla            4.000000
Chehlla Bandi      6.000000
Chella Bandi       5.000000
Gojra              5.588235
Jalalabad          7.000000
Madina Market      3.000000
Naloochi           8.000000
Plate              5.090909
Tanga Stand        6.000000
Tarqabad           5.666667
Name: No of Rooms, dtype: float64

The location with the most spacious homes is Naloochi with an average of 8.0 rooms.


## Key insights:
#### Most Spacious Location: 
The location with the most spacious homes is Naloochi, with an average of 8.0 rooms, indicating larger houses compared to other areas.
#### Close Competitors: 
Locations such as Bela Noor Shah (7.5) and Jalalabad (7.0) also have relatively spacious homes.
#### Least Spacious Locations: 
Areas like Balapeer (3.0) and Madina Market (3.0) tend to have smaller homes with fewer rooms.
#### Variation in Room Averages: 
The average number of rooms varies significantly across locations, reflecting diverse housing characteristics in the dataset.

3- What is the most common condition of houses (e.g., Good, Fair, Excellent) in the data? Does the majority align with higher or lower prices?

In [36]:
# Find the most common condition (mode)
most_common_condition = house_data['Condition '].mode()[0]

# Group by condition and calculate the average house price
average_price_by_condition = house_data.groupby('Condition ')['House Prices (Millions)'].mean()

# Identify whether the most common condition aligns with higher or lower prices
common_condition_price = average_price_by_condition[most_common_condition]
overall_average_price = house_data['House Prices (Millions)'].mean()
alignment = "higher" if common_condition_price > overall_average_price else "lower"

# Print results
print(f" -The most common condition of houses is '{most_common_condition}'.")
print(f" -Average house price for '{most_common_condition}' condition: {common_condition_price:.2f} million.")
print(f" -Overall average house price: {overall_average_price:.2f} million.")
print(f" -The majority condition aligns with {alignment} prices.")


 -The most common condition of houses is 'Good'.
 -Average house price for 'Good' condition: 12.95 million.
 -Overall average house price: 12.00 million.
 -The majority condition aligns with higher prices.


## key insights:
#### Most Common Condition:
The most common condition of houses in the dataset is 'Good', indicating that a majority of the houses fall under this category.
#### Higher Prices: 
Houses in 'Good' condition have an average price of 12.95 million, which is slightly higher than the overall average price of 12.00 million.
#### Market Alignment: 
The majority condition aligns with higher prices, suggesting that 'Good' condition homes are generally valued above average in the dataset.
#### Condition-Price Correlation: 
This trend highlights the positive correlation between house condition and price, with better conditions commanding higher values.

 ## Part 2: Dispersion (Variability in Data)

1- Which location shows the greatest variability in house prices? Use measures such as range, interquartile range (IQR), or standard deviation to support your conclusion.

In [37]:
# Group by location
grouped = house_data.groupby('Location')['House Prices (Millions)']

# Calculate range, standard deviation, and IQR
ranges = grouped.max() - grouped.min()
std_devs = grouped.std()
iqrs = grouped.apply(lambda x: np.percentile(x, 75) - np.percentile(x, 25))

# Combine results into a single DataFrame
metrics = pd.DataFrame({
    "Location": ranges.index,
    "range": ranges.values,
    "std_dev": std_devs.values,
    "iqr": iqrs.values
})

# Find maximum variability
max_range = metrics.loc[metrics['range'].idxmax()]
max_std_dev = metrics.loc[metrics['std_dev'].idxmax()]
max_iqr = metrics.loc[metrics['iqr'].idxmax()]

# Display results
print("Variability by Location:")
print(metrics)

print(f"\nGreatest Variability by Range: {max_range['Location']} ({max_range['range']} million)")
print(f"Greatest Variability by Std Dev: {max_std_dev['Location']} ({max_std_dev['std_dev']:.2f} million)")
print(f"Greatest Variability by IQR: {max_iqr['Location']} ({max_iqr['iqr']} million)")


Variability by Location:
           Location  range   std_dev   iqr
0            Ambore   23.0  7.446103  1.00
1         Balapeer     0.0       NaN  0.00
2    Bela Noor Shah   20.0  9.604686  7.25
3    Bela noor shah    8.0  4.358899  4.00
4   Bella Noor Shah    3.0  1.732051  1.50
5           Chatter    8.0  3.011091  4.50
6           Chehlla    7.0  4.949747  3.50
7     Chehlla Bandi   13.0  5.560276  4.75
8      Chella Bandi    0.0       NaN  0.00
9             Gojra   18.0  4.846921  6.00
10        Jalalabad   11.0  7.778175  5.50
11    Madina Market    0.0       NaN  0.00
12         Naloochi   14.0  9.899495  7.00
13            Plate   24.0  6.768913  6.00
14      Tanga Stand    0.0       NaN  0.00
15         Tarqabad    2.0  1.000000  1.00

Greatest Variability by Range: Plate (24.0 million)
Greatest Variability by Std Dev: Naloochi (9.90 million)
Greatest Variability by IQR: Bela Noor Shah (7.25 million)


## Conclusion:
Plate shows the greatest range (24 million), but this may be influenced by an outlier. Naloochi has the highest standard deviation (9.90 million), and Bela Noor Shah has the highest IQR (7.25 million), indicating more consistent variability.

## Key Insight:
Plate has the highest range, but Naloochi and Bela Noor Shah show greater consistent variability by standard deviation and IQR, respectively.

2- Analyze the variability in rent prices across locations. Are rents more stable in some locations compared to others?

In [38]:
# Assuming house_data includes rent prices in the 'Rent Prices (Millions)' column
grouped_rent = house_data.groupby('Location')['Rent Prices']

# Calculate range, standard deviation, and IQR for rent prices
rent_ranges = grouped_rent.max() - grouped_rent.min()
rent_std_devs = grouped_rent.std()
rent_iqrs = grouped_rent.apply(lambda x: np.percentile(x, 75) - np.percentile(x, 25))

# Combine results into a single DataFrame for rent prices
rent_variability_metrics = pd.DataFrame({
    "Location": rent_ranges.index,
    "Range": rent_ranges.values,
    "Standard Deviation": rent_std_devs.values,
    "IQR": rent_iqrs.values
})

# Display results
print("Rent Variability by Location:")
print(rent_variability_metrics)

Rent Variability by Location:
           Location   Range  Standard Deviation      IQR
0            Ambore   30000        12018.504252  20000.0
1         Balapeer        0                 NaN      0.0
2    Bela Noor Shah   22000        10436.314803   9250.0
3    Bela noor shah  190000        98488.578018  95000.0
4   Bella Noor Shah    5000         2886.751346   2500.0
5           Chatter   30000         9267.026372   6750.0
6           Chehlla   30000        21213.203436  15000.0
7     Chehlla Bandi   15000         6454.972244   7500.0
8      Chella Bandi       0                 NaN      0.0
9             Gojra  297000        67952.892420  10000.0
10        Jalalabad   60000        42426.406871  30000.0
11    Madina Market       0                 NaN      0.0
12         Naloochi    5000         3535.533906   2500.0
13            Plate   35000         9234.323325   3500.0
14      Tanga Stand       0                 NaN      0.0
15         Tarqabad    8000         4358.898944   4000.0


## key insights:
Stable Rent Locations: Balapeer, Chella Bandi, Madina Market, and Tanga Stand show low or no variability in rent prices, indicating stability.
Variable Rent Locations: Bela noor shah and Gojra have high variability in rent prices, with large ranges, standard deviations, and IQRs.

3- Compare the variability in house prices for houses with and without a garden. Do gardens significantly influence the consistency of pricing?

In [39]:

# Group by garden status and calculate range, standard deviation, and IQR for house prices
grouped_garden = house_data.groupby('Garden')['House Prices (Millions)']

# Calculate range, standard deviation, and IQR for house prices
price_ranges = grouped_garden.max() - grouped_garden.min()
price_std_devs = grouped_garden.std()
price_iqrs = grouped_garden.apply(lambda x: np.percentile(x, 75) - np.percentile(x, 25))

# Combine results into a single DataFrame for house prices with and without gardens
price_variability_metrics = pd.DataFrame({
    "Garden": price_ranges.index,
    "Range": price_ranges.values,
    "Standard Deviation": price_std_devs.values,
    "IQR": price_iqrs.values
})

# Display results
print("House Price Variability by Garden Status:")
print(price_variability_metrics)


House Price Variability by Garden Status:
  Garden  Range  Standard Deviation   IQR
0     No   14.0            3.682742  2.00
1    Yes   23.0            6.404549  6.25


## Key Insights:

#### Houses Without a Garden:
Houses without a garden show relatively lower variability in prices, with a smaller IQR and standard deviation, indicating more consistency in pricing.

#### Houses With a Garden:
Houses with a garden exhibit higher variability, with a larger range, higher standard deviation, and a bigger IQR, suggesting more fluctuation in pricing.

#### Conclusion:
Houses with a garden tend to have greater variability in prices compared to houses without a garden. The larger range, standard deviation, and IQR for houses with gardens indicate that gardens significantly influence the consistency of house prices, making them more variable.

## Part 3: Correlation (Relationships in Data)

1- Is there a relationship between the size of a house (area) and its house price? Use correlation analysis to identify whether larger houses are  priced higher.

In [40]:
# Assuming 'house_data' contains the columns 'Area' for size and 'House Price' for price
correlation = house_data['Area (Marla)'].corr(house_data['House Prices (Millions)'])

print(f"Correlation between House Area and House Price: {correlation}")

Correlation between House Area and House Price: 0.6660582581828108


## Key insights:
The correlation coefficient of 0.67 indicates a moderate positive relationship between house area and house price. This suggests that, on average, larger houses tend to have higher prices, although other factors may also influence the price. While the correlation is not perfect, there is a noticeable trend where increasing house size is associated with an increase in price.

2-Investigate whether the number of rooms or the number of washrooms has a stronger correlation with rent prices.

In [41]:
# Calculate the correlation coefficients
corr_rooms_rent = house_data['No of Rooms'].corr(house_data['Rent Prices'])
corr_washrooms_rent = house_data['No of Washrooms'].corr(house_data['Rent Prices'])

# Display the results
print(f"Correlation between Number of Rooms and Rent Prices: {corr_rooms_rent}")
print(f"Correlation between Number of Washrooms and Rent Prices: {corr_washrooms_rent}")

# Based on these results, we can interpret which feature has a stronger relationship with rent prices.

Correlation between Number of Rooms and Rent Prices: 0.2611970960304241
Correlation between Number of Washrooms and Rent Prices: 0.27070307991104675


## Key insights:
The number of washrooms has a slightly stronger positive correlation with rent prices compared to the number of rooms. While both correlations are relatively weak, the data suggests that the number of washrooms might be a slightly more influential factor in determining rent prices than the number of rooms.

3- Does the year of construction correlate with house prices? For example, are newer houses priced higher, or is the relationship weak?

In [42]:

correlation = house_data['Year of Construction'].corr(house_data['House Prices (Millions)'])

# Print the correlation result
print("Correlation between Year of Construction and House Price:", correlation)


Correlation between Year of Construction and House Price: -0.07279394977179576


## Key insights:
The correlation between the year of construction and house price is -0.0728, indicating a very weak negative relationship. This suggests that the age of the house has little to no effect on its price, and factors like location, size, and condition are more likely to influence pricing.